src/org/sosy_lab/java_smt/basicimpl/AbstractArrayFormulaManager.java

```diff
@@ -39,8 +39,11 @@ public <TI extends Formula, TE extends Formula> TE select(
     //    Rational
 
     final TFormulaInfo term = select(extractInfo(pArray), extractInfo(pIndex));
-
-    return (TE) getFormulaCreator().encapsulate(elementType, term);
+    TE result = (TE) getFormulaCreator().encapsulate(elementType, term);
+    if (Generator.isLoggingEnabled()) {
+      ArrayGenerator.logSelect(result, pArray, pIndex);
```

Comment:
Logging for the SMTLIB2 generator is done using the same global generator objects. This is an issue when more than one solver instance tries to output SMTLIB2 code at the same time.
We should consider reimplementing this using a delegate. At  the very least the logged SMTLIB2 output needs to be linked to the SolverContext that created it.